### PR TITLE
[WIP] - x509 cache hint for registration entries

### PR DESCRIPTION
**Pull request by ValFadeev**
This is a draft PR aiming to implement some of the suggestions discussed in https://github.com/spiffe/spire/issues/6129. It adds a new field to `RegisteredEntry` represented by an opaque protobuf-encoded message. The message is meant to contain flags and other fields guiding the agent's behavior with respect to pre-fetching and caching of X509 SVIDs. The intention is to reduce unnecessary work performed by the agent, in case the client is unlikely to ever ask for an X509 SVID and prefers JWT instead. For now only one boolean flag `JWTOnly` is introduced. Other fields may be added later to fine-tune the logic and cover various edge-cases. 
The bulk of this draft is to define the new field in the message definition, update the datastore and have the existing tests pass. The actual usage of the field in `fetchEntries` is purely indicative and requires further consideration.

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [ ] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**

* adds a new field to `RegisteredEntry` and the underlying datastore table
* allows flagging registration entries as preferring JWT SVID over X509
**Description of change**

**Which issue this PR fixes**

Fixes https://github.com/spiffe/spire/issues/6129

### CODE REVIEW

File: proto/spire/common/common.proto [hunk 1]
```diff
@@ -96,6 +96,14 @@ message RegistrationEntry {
...
     /** Time of creation, in seconds from epoch */
     int64 created_at = 15;
+    /** Represents a set of options informing the agent behaviour with respect to
+    pre-fecthing and caching X509 SVIDs. This is meant to prevent unnecessary effort
+    spent on providing X509 SVIDs to workloads, which are likely to ask only for JWT SVID. */
+    message X509SvidCacheHint {
+        /** Flag indicating whether the workload is likely to require only JWT SVID. */
+        bool jwt_only = 1;
+    }
+    X509SvidCacheHint x509_svid_cache_hint = 16;
```
Comment (sorindumitru) [inline proto/spire/common/common.proto]: We discussed this today and we think this would be more useful if it was a generic flag field that we can use in the future for non x509-svid cache flags.

For the flag itself, we think `disable_x509_svid_prefetch` would be better since it allows us to specify more easily which SVID types should be prefetched. This might come in handy if we get a new SVID type, for example WIT-SVID.

**Comment by amartinezfayo** [comment]
Hi @ValFadeev, are you planning to go back to this PR?

**Comment by ValFadeev** [comment]
Hi @amartinezfayo, apologies for the delay. I had a few conflicting priorities lately, but I do plan to pick this back up over the coming week.

**Comment by ValFadeev** [comment]
@amartinezfayo offering this for another review. I am not sure I implemented the semantics of an optional field the right way when [loading from row](https://github.com/spiffe/spire/pull/6360/files#diff-a972d23257a60257623830219f33a521a3c95b8223c6e3f80d840d699aa81d3dR3963). I also understand this change may have to be split in two, with the schema migration going first. But hopefully, this gives an outline of the functionality, as it was last described.

File: pkg/agent/manager/sync.go [hunk 2]
```diff
@@ -345,7 +345,9 @@ func (m *manager) fetchEntries(ctx context.Context) (_ *cache.UpdateEntries, _ *
 		case entry.StoreSvid:
 			storeEntries[entryID] = entry
 		default:
-			cacheEntries[entryID] = entry
+			if !entry.CacheHintFlags.GetDisableX509SvidPrefetch() {
+				cacheEntries[entryID] = entry
+			}
```
Comment (sorindumitru) [inline pkg/agent/manager/sync.go]: This would mean that the entry is not added to the cache, which means that no workload would be able to request a SVID for it. I think the flag should end up in the lru cache where it can inform if the entry needs to prefetched or not but still allow a workload to request an x509-svid if it so desires.

I think [syncSVIDsWithSubscribers](https://github.com/spiffe/spire/blob/5512c9bbb7f167b0e56d0b6fbfb2cbd2e69ecebb/pkg/agent/manager/cache/lru_cache.go#L781) might be the place to do this.

File: pkg/server/datastore/sqlstore/sqlstore.go [hunk 3]
```diff
@@ -3944,6 +3960,15 @@ func fillEntryFromRow(entry *common.RegistrationEntry, r *entryRow) error {
 		entry.CreatedAt = roundedInSecondsUnix(r.CreatedAt.Time)
 	}
 
+	if r.CacheHintFlags.Valid {
+		if len(r.CacheHintFlags.V) > 0 {
+			entry.CacheHintFlags = &common.RegistrationEntry_CacheHintFlags{}
+			if err := proto.Unmarshal(r.CacheHintFlags.V, entry.CacheHintFlags); err != nil {
+				return newSQLError("invalid value for X.509 cache hint: %s", err)
```
Comment (sorindumitru) [inline pkg/server/datastore/sqlstore/sqlstore.go]:
```suggestion
				return newSQLError("could not parse cache hint flags: %s", err)
```

**Comment by ValFadeev** [comment]
@srikalyan mind if I bring this up to date and actually add the proposed logic making use of the flag, with the deadline by the end of the week. Then we can see how the work can be potentially reused/generalised. Thanks.

**Comment by srikalyan** [comment]
@ValFadeev please go ahead and let me know if I can help in anyway.

**Comment by sorindumitru** [comment]
I think you'll also need to make changes to the Entry type in [spire-api-sdk](https://github.com/spiffe/spire-api-sdk/blob/main/proto/spire/api/types/entry.proto) for the agent to get and be able to use the new field.

File: proto/spire/common/common.proto [hunk 4]
```diff
@@ -99,6 +99,15 @@ message RegistrationEntry {
     string hint = 14;
     /** Time of creation, in seconds from epoch */
     int64 created_at = 15;
+    /** Represents a set of flags informing the agent behaviour with respect to
+    pre-fecthing and caching SVIDs. This is meant to prevent unnecessary effort
+    spent on generating SVIDs of types, which are unlikely to be needed. */
+    message CacheHintFlags {
```
Comment (sorindumitru) [inline proto/spire/common/common.proto]: Since we're likely going to use this even for non-flag fields, could we call it something else? Nothing better than `OptionalFields` comes to me right now, let me know if you have other suggestions.

File: pkg/server/datastore/sqlstore/models.go [hunk 5]
```diff
@@ -100,6 +100,10 @@ type RegisteredEntry struct {
 
 	// TTL of JWT identities derived from this entry
 	JWTSvidTTL int32 `gorm:"column:jwt_svid_ttl"`
+
+	// CacheHintFlags contains a set of options and flags that inform the
+	// agent behaviour with respect to pre-fetching and refreshing X509 SVIDs
+	CacheHintFlags []byte `gorm:"size:255,column:cache_hint_flags"` // corresponds to TINYBLOB in MySQL
```
Comment (sorindumitru) [inline pkg/server/datastore/sqlstore/models.go]: Since we may end up having other bigger fields in there, maybe we can go one size up to have some more space? I think that would be `size:65535`

File: proto/spire/common/common.proto (hunk 4 above)
Comment (ValFadeev) [reply proto/spire/common/common.proto]: I would say something like AdditionalAttributes. This is to distinguish them from optional fields in the containing struct, such as Expiry.

File: pkg/server/datastore/sqlstore/migration.go [hunk 6]
```diff
@@ -517,8 +517,7 @@ func migrateVersion(tx *gorm.DB, currVersion int, log logrus.FieldLogger) (versi
 }
 
 func migrateToV24(tx *gorm.DB) error {
-	// Add agent_version column to attested_node_entries table
-	if err := tx.AutoMigrate(&AttestedNode{}).Error; err != nil {
+	if err := tx.AutoMigrate(&AttestedNode{}, &RegisteredEntry{}).Error; err != nil {
```
Comment (ValFadeev) [inline pkg/server/datastore/sqlstore/migration.go]: Since there is currently no released version between my changes and https://github.com/spiffe/spire/pull/6542, I am temporarily combining migrations here to have tests pass. Once 1.14.2 is released, I will only leave my table and increment the version.

Comment (sorindumitru) [reply pkg/server/datastore/sqlstore/migration.go]: Usually we seemed to have created independent migrations for different data store changes. I think they're not strictly speaking needed, but they might be useful for the case of local development.

**Review by sorindumitru — COMMENTED** [review]
Thanks for these changes, @ValFadeev. Are you also planning on adding support for this to the CLI?

File: cmd/spire-server/cli/entry/create.go [hunk 7]
```diff
@@ -209,8 +214,12 @@ func (c *createCommand) parseConfig() ([]*types.Entry, error) {
 
 		selectors = append(selectors, cs)
 	}
-
 	e.Selectors = selectors
+
+	additionalAttributes := &types.Entry_AdditionalAttributes{}
+	additionalAttributes.DisableX509SvidPrefetch = c.disableX509SVIDPrefetch
+	e.AdditionalAttributes = additionalAttributes
```
Comment (sorindumitru) [inline cmd/spire-server/cli/entry/create.go]: Could we only create the additional attributes if one of the attributes in it is defined? Would avoid having to parse the whole field.

File: pkg/server/api/entry.go [hunk 8]
```diff
@@ -179,23 +187,35 @@ func RegistrationEntryToProto(e *common.RegistrationEntry) (*types.Entry, error)
...
+	return entry, nil
+}
+
+func ProtoFromAdditionalAttributes(in *common.RegistrationEntry_AdditionalAttributes) *types.Entry_AdditionalAttributes {
+	if in != nil {
+		return &types.Entry_AdditionalAttributes{
+			DisableX509SvidPrefetch: in.DisableX509SvidPrefetch,
+		}
+	}
+	return nil
```
Comment (sorindumitru) [inline pkg/server/api/entry.go:222]: I think we would also need a similar item for doing the reverse, and it should be used in `ProtoToRegistrationEntryWithMask`.

File: pkg/server/datastore/sqlstore/migration.go (hunk 6 above)
Comment (sorindumitru) [reply pkg/server/datastore/sqlstore/migration.go]: This still needs to be addressed.

File: pkg/server/datastore/sqlstore/migration.go [hunk 9]
```diff
@@ -506,6 +506,8 @@ func migrateVersion(tx *gorm.DB, currVersion int, log logrus.FieldLogger) (versi
 	switch currVersion {
 	case 23:
 		err = migrateToV24(tx)
```
Comment (sorindumitru) [inline pkg/server/datastore/sqlstore/migration.go:511]: We'll need to do both of them if we're on 23, so maybe:
```suggestion
		err = migrateToV24(tx)
		fallthrough
```

**Review by sorindumitru — COMMENTED** [review]
Didn't get a chance to look through all of it yet, so there may be more comments coming.

File: pkg/server/api/entry.go [hunk 10]
```diff
@@ -179,23 +187,49 @@ func RegistrationEntryToProto(e *common.RegistrationEntry) (*types.Entry, error)
...
+		Admin:                e.Admin,
+		Downstream:           e.Downstream,
+		ExpiresAt:            e.EntryExpiry,
+		DnsNames:             slices.Clone(e.DnsNames),
+		RevisionNumber:       e.RevisionNumber,
+		StoreSvid:            e.StoreSvid,
+		JwtSvidTtl:           e.JwtSvidTtl,
+		Hint:                 e.Hint,
+		CreatedAt:            e.CreatedAt,
+		AdditionalAttributes: ProtoFromAdditionalAttributes(e.AdditionalAttributes),
```
Comment (sorindumitru) [inline pkg/server/api/entry.go]: Looks like we set the `AdditionalAttributes` twice. Once here and once in a few lines down.

**Review by sorindumitru — COMMENTED** [review]
Thanks @ValFadeev, this is shaping up nicely. Only one small comment.

**Comment by sorindumitru** [comment]
Hi @ValFadeev , sorry for the delays here. We've merged the API changes, would you mind repointing the dependency to the `github.com/spiffe/spire-api-sdk@next`?

File: cmd/spire-server/cli/entry/update.go [hunk 11]
```diff
@@ -204,6 +209,13 @@ func (c *updateCommand) parseConfig() ([]*types.Entry, error) {
 	}
 
 	e.Selectors = selectors
+
+	if c.disableX509SVIDPrefetch {
```
Comment (sorindumitru) [inline cmd/spire-server/cli/entry/update.go:225]: I think this makes it impossible to remove the flag once set. We need a way to determine if the flag was set or not. Maybe we can use BoolFunc so that we know if it was set or not? We could even have a more general bool "setAdditionalAttributes" which is set to true if the BoolFunc is called?

File: pkg/server/datastore/sqlstore/models.go [hunk 12]
```diff
@@ -100,6 +100,11 @@ type RegisteredEntry struct {
 
 	// TTL of JWT identities derived from this entry
 	JWTSvidTTL int32 `gorm:"column:jwt_svid_ttl"`
+
+	// AdditionalAttributes may contain a number of optional fields controlling
+	// the various aspects of the agent's behaviour with respect to a given
+	// registration entry
+	AdditionalAttributes []byte `gorm:"size:65535,column:additional_attributes"`
```
Comment (sorindumitru) [inline pkg/server/datastore/sqlstore/models.go:107]: Maybe we could also enforce this in the data store. It would at least make for better error messages, if nothing else.

File: cmd/spire-server/cli/entry/update.go (hunk 11 above)
Comment (ValFadeev) [reply cmd/spire-server/cli/entry/update.go:225]: That is not what I observe - this is run on a clean build of the server:
```
./bin/spire-server entry show -entryID prefetch-disabled -socketPath=/tmp/spire-server/private/api.sock
Found 1 entry
Entry ID                : prefetch-disabled
SPIFFE ID               : spiffe://example.org/workload-with-prefetch-disabled
Parent ID               : spiffe://example.org/cluster/test
Revision                : 6
X509-SVID TTL           : default
JWT-SVID TTL            : default
Selector                : unix:uid:1004
DisableX509SvidPrefetch : true
```
Now if an update performed without the flag set it is cleared in the entry:
```
./bin/spire-server entry update -entryID prefetch-disabled -selector "unix:uid:1004" -parentID spiffe://example.org/cluster/test -spiffeID spiffe://example.org/workload-with-prefetch-disabled -socketPath=/tmp/spire-server/private/api.sock
Entry ID                : prefetch-disabled
SPIFFE ID               : spiffe://example.org/workload-with-prefetch-disabled
Parent ID               : spiffe://example.org/cluster/test
Revision                : 7
X509-SVID TTL           : default
JWT-SVID TTL            : default
Selector                : unix:uid:1004
```
and set again:
```
./bin/spire-server entry update -entryID prefetch-disabled -selector "unix:uid:1004" -parentID spiffe://example.org/cluster/test -spiffeID spiffe://example.org/workload-with-prefetch-disabled -disableX509SVIDPrefetch -socketPath=/tmp/spire-server/private/api.sock
Entry ID                : prefetch-disabled
SPIFFE ID               : spiffe://example.org/workload-with-prefetch-disabled
Parent ID               : spiffe://example.org/cluster/test
Revision                : 8
X509-SVID TTL           : default
JWT-SVID TTL            : default
Selector                : unix:uid:1004
DisableX509SvidPrefetch : true
```
Tha said, it could indeed be useful to introduce an overarching bool, such as "setAdditionalAttributes", to indicate whether any one of the additional attributes has been provided, to set it up for generalisation. I suppose the flag definition could similar to the below one (for every new additional attribute implemented"
```go
	f.BoolFunc("disableX509SVIDPrefetch", "A boolean value that, when set, disables prefetching X509 SVID for this entry", 
	func(_ string) error {
		c.additionalAttributesSet = true
		c.disableX509SVIDPrefetch = true
		return nil
	})
```
So that the entry field would be populated like this
```go
	if c.additionalAttributesSet {
		e.AdditionalAttributes = &types.Entry_AdditionalAttributes{}
	}

	if c.disableX509SVIDPrefetch {
		e.AdditionalAttributes.DisableX509SvidPrefetch = true
	}
```
If this makes sense, I can make the adjustment.

File: pkg/server/datastore/sqlstore/migration.go [hunk 13]
```diff
@@ -509,6 +509,12 @@ func migrateVersion(tx *gorm.DB, currVersion int, log logrus.FieldLogger) (versi
...
+		if err != nil {
+			break
+		}
+		fallthrough
```
Comment (sorindumitru) [inline pkg/server/datastore/sqlstore/migration.go]: Since version 24 was release in 1.14.3/1.14.4, we no longer need to fallthrough. Sorry about that, should have realised that would be the case when I asked you to do this.

File: pkg/server/datastore/sqlstore/sqlstore.go [hunk 14]
```diff
@@ -4497,6 +4532,22 @@ func modelToBundle(model *Bundle) (*common.Bundle, error) {
 	return bundle, nil
 }
 
+func validateAditionalAttributes(additionalAttributes *common.RegistrationEntry_AdditionalAttributes) error {
+	if additionalAttributes == nil {
+		return nil
+	}
+
+	marshaledAdditionalAttributes, err := proto.Marshal(additionalAttributes)
```
Comment (sorindumitru) [inline pkg/server/datastore/sqlstore/sqlstore.go:4540]: This looks ok, but it means we marsh the attributes twice on update/creation. Could we instead apply these validations at that point? Maybe we can make this function to also return the bytes.

**Comment by kfox1111** [comment]
I'm curious. would the AdditionalAttributes support in this pr make it easier to add https://github.com/spiffe/spire/issues/6069
